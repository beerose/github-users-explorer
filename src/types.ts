import * as github from "./githubClient";

export type User = Pick<
  github.User,
  | "avatar_url"
  | "bio"
  | "blog"
  | "company"
  | "html_url"
  | "location"
  | "login"
  | "name"
>;

export type Repository = Pick<
  github.Repository,
  "html_url" | "description" | "stargazers_count" | "name"
>;
