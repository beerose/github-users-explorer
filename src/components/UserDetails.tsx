/** @jsx jsx */
import { css, jsx } from "@emotion/core";
import { ComponentProps } from "react";
import { transparentize } from "polished";

import { prefixWithHttps } from "../utils";
import { styled, useTheme } from "../theme";

import { RepositoryDetails } from "./RepositoryDetails";
import { User, Repository } from "../types";

const UserInfoList = "ul";
const UserInfoListItem: React.FC<ComponentProps<"li">> = props =>
  props.children ? <li {...props} /> : null;

interface UserWebsiteLinkProps
  extends Omit<ComponentProps<"a">, "href" | "children"> {
  url: string;
}
const UserWebsiteLink = ({ url, ...rest }: UserWebsiteLinkProps) => {
  const { colors } = useTheme();

  return url ? (
    <a
      href={prefixWithHttps(url)}
      css={css`
        color: ${colors.primary};
      `}
      {...rest}
    >
      {url}
    </a>
  ) : null;
};

const Avatar = styled.img`
  border-radius: 50%;
  width: 8em;
`;

const UserDetailsSection = styled.section`
  display: flex;
  align-items: center;
  flex-direction: column;
  margin: 2em;
`;

const UserDetailsHeader = styled.header`
  text-align: center;

  h2 {
    margin-bottom: 0.2em;
  }
`;

const UserLogin = styled.a`
  text-decoration: none;

  color: ${props => transparentize(0.5, props.theme.colors.text)};

  ::before {
    content: "@";
  }
`;

const Bio = styled.p`
  margin: 1.5em;
  max-width: 20em;
  text-align: center;
`;

const Repositories = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

interface UserDetailsProps {
  user: User;
  repos: Repository[];
}
export const UserDetails = ({ user, repos }: UserDetailsProps) => {
  return (
    <UserDetailsSection>
      <Avatar src={user.avatar_url} alt="User avatar" />
      <UserDetailsHeader>
        <h2>{user.name}</h2>
        <UserLogin href={user.html_url}>{user.login}</UserLogin>
      </UserDetailsHeader>
      <Bio>{user.bio}</Bio>
      <UserInfoList>
        <UserInfoListItem>{user.company}</UserInfoListItem>
        <UserInfoListItem>{user.location}</UserInfoListItem>
        <UserInfoListItem>
          <UserWebsiteLink url={user.blog} />
        </UserInfoListItem>
      </UserInfoList>
      <Repositories>
        {repos.map(repo => (
          <RepositoryDetails key={repo.name} repository={repo} />
        ))}
      </Repositories>
    </UserDetailsSection>
  );
};
