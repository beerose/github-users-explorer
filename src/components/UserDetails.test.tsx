import React from "react";
import "@testing-library/jest-dom/extend-expect";

import { render } from "../../test/testUtils";
import { User } from "../types";

import { UserDetails } from "./UserDetails";

const user: User = {
  name: "Anders Hejlsberg",
  login: "ahejlsberg",
  avatar_url: "https://avatars0.githubusercontent.com/u/4226954?s=400&v=4",
  html_url: "https://github.com/ahejlsberg",
  company: "Microsoft",
  bio:
    "Microsoft Technical Fellow and lead architect of TypeScript. Original designer of C#, Delphi, and Turbo Pascal.",
  blog: "",
  location: "Redmond, WA, USA",
};

describe("UserDetails", () => {
  it("displays user name in a heading", () => {
    const { getByText } = render(<UserDetails user={user} />);

    expect(getByText(user.name)).toBeInstanceOf(HTMLHeadingElement);
  });

  it("login is an anchor to GitHub account", () => {
    const { getByText } = render(<UserDetails user={user} />);

    expect(getByText(user.login)).toHaveAttribute("href");
  });

  it("matches snapshot", () => {
    const { asFragment } = render(<UserDetails user={user} />);

    expect(asFragment()).toMatchSnapshot();
  });
});
