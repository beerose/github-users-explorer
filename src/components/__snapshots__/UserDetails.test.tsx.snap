// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDetails matches snapshot 1`] = `
<DocumentFragment>
  <section
    class="css-chsl2o"
  >
    <img
      alt="User avatar"
      class="css-8jrh0l"
      src="https://avatars0.githubusercontent.com/u/4226954?s=400&v=4"
    />
    <header
      class="css-1i6qyll"
    >
      <h2>
        Anders Hejlsberg
      </h2>
      <a
        class="css-1ceuyr5"
        href="https://github.com/ahejlsberg"
      >
        ahejlsberg
      </a>
    </header>
    <p
      class="css-g6ecle"
    >
      Microsoft Technical Fellow and lead architect of TypeScript. Original designer of C#, Delphi, and Turbo Pascal.
    </p>
    <ul>
      <li>
        Microsoft
      </li>
      <li>
        Redmond, WA, USA
      </li>
      <li />
    </ul>
    <div
      class="css-rtde4j"
    >
      <a
        class="css-5r94d3"
        href="https://github.com/ahejlsberg/typescript-build2016-demos"
      >
        <div
          class="css-1p91yv2"
        >
          <svg
            fill="none"
            height="24"
            stroke="currentColor"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
            viewBox="0 0 24 24"
            width="24"
          >
            <path
              d="M12 2L15.09 8.26 22 9.27 17 14.14 18.18 21.02 12 17.77 5.82 21.02 7 14.14 2 9.27 8.91 8.26 12 2z"
            />
          </svg>
          64
        </div>
        <h3>
          typescript-build2016-demos
        </h3>
        <p>
          Collection of TypeScript demo projects form //build 2016
        </p>
      </a>
    </div>
  </section>
</DocumentFragment>
`;
